Drop support for Python 2.7!

zbruick opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/Unidata/MetPy/blob/master/CONTRIBUTING.md
-->

#### Description Of Changes
What do you do when it's a Friday afternoon? Drop 2.7! 

In all seriousness, according to our docs, we're dropping 2.7 in "Fall 2019". Given that meteorological (real) fall is here and astronomical fall is 8 days away, this is likely for sure going in 0.12. (Not sure when 0.11 will get out the door, but not sure if we should include this in that or not). 

I'm sure I missed some stuff, so feel free to push commits to this, or review and comment if you have small items to address. There is one open item on #801 that I didn't fix yet, as the comment was vague and I didn't have time to look into it yet. 

Whenever 2.7 is dropped, celebration is in order!

#### Checklist
<!--
Feel free to remove check-list items aren't relevant to your change

Please use keywords (e.g., Fixes, Closes) to create link to the issues or pull
requests you resolved, so that they will automatically be closed when your pull
request is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
- [x] Closes #801
- [x] Fully documented

dopplershift commented on metpy/__init__.py:
```diff
@@ -16,9 +16,5 @@
-except AttributeError:
-    # Add our own for MetPy on Python 2.7
-    logging.getLogger(__name__).addHandler(logging.StreamHandler())
+# Added in Python 3.2, will log anything warning or higher to stderr
```
Should be able to just nuke this entirely. Original code was just checking for this attribute and then doing more on 2.7 (where it didn't exist).

dopplershift commented on metpy/cbook.py:
```diff
@@ -16,12 +16,6 @@
     string_type = str
```
Should be able to nuke this block too

dopplershift commented on docs/installguide.rst:
```diff
@@ -2,47 +2,19 @@
-* matplotlib >= 2.0.0
-* numpy >= 1.12.0
-* scipy >= 0.17.0
+* matplotlib >= 2.1.0
```
Are these just generic "time to bump versions" updates?

zbruick replied:
Yes they are

zbruick replied:
Figuring that this PR will be in 0.12 later this fall. Some of these bumps maybe should occur in 0.11?

dopplershift replied:
If there's not a *need*, I'd rather not add them to the list for this release.

dopplershift commented on metpy/io/_tools.py:
```diff
@@ -14,14 +12,7 @@
-    bytearray_to_buff = buffer
-except NameError:
-    bytearray_to_buff = memoryview
+bytearray_to_buff = memoryview
```
Can we just replace all calls to `bytearray_to_buff` with `memoryview` instead?

zbruick replied:
Yup that works

dopplershift commented on metpy/plots/wx_symbols.py:
```diff
@@ -17,12 +17,7 @@
-    code_point = unichr
-except NameError:
-    code_point = chr
+code_point = chr
```
Should be able to just change all uses of `code_point` to `chr` directly.

dopplershift commented:
One good one to do to modernize the code base is to convert all uses of `super` to just `super()`.

dopplershift commented:
Actually, before doing that, it might be good to see what we can do about having flake8 catch that, and other modernizations, that we make.

dopplershift commented:
Another one to add is updating `setup.cfg` to remove dependencies (I'm sure you had before I rewrote the build system) as well as remove the `bdist_wheel` section to no longer say we build universal wheels.

dopplershift commented:
Other things coming to mind:
- In several places we have code that does some hacking around the crappy function signature support on Python 2.7. We can clean that up.
- A lot of places using `kwargs.pop` should probably be changed to use kw-only arguments

zbruick commented:
> Actually, before doing that, it might be good to see what we can do about having flake8 catch that, and other modernizations, that we make.

Have been looking for any plugins and haven't seen anything useful yet. Lots of py2 to py 2/3 support online. Doesn't really look like there are any good linters for conversion of py2/3 to py3. Looking at astropy, they suggested running coverage just with py2 and py3 separately to see what was only covered in either to remove dead code. Is that something that can easily be done here?

zbruick commented:
Ok I've cleaned up setup.cfg and removed uses of `kwargs.pop` wherever possible (only remaining ones are within the plot module).

zbruick commented:
> * In several places we have code that does some hacking around the crappy function signature support on Python 2.7. We can clean that up.

Clarification on where these hacks exist (or what they look like) would be great

dopplershift commented:
I’ll look at the signature stuff. We should be able to look at codecov and see what it says about new uncovered lines with this PR to find other blocks of unneeded code (assuming it’s working).

zbruick commented:
Found one thing in `units.py` that could be removed. There is also a comment at the start of the import section of `grid.py` to update when 2.7 is removed, but it isn't readily clear what should be done there.

I believe the only things that were mentioned that haven't been done is to update `super()` and clean up hacks.

zbruick commented:
Fixed imports in grid.py and tools.py.

Added errors to setup.py and __init__.py if Python 2 is detected to either fail to install or fail to import. Verified that the error message in setup.py is returned in `pip install -e.` is executed in a Py2 environment.

zbruick commented:
Took care of super() with the last commit. All that is now left is any function signature updates that are needed.

dopplershift commented:
Another one is we can remove is inheriting from `object`.

So while there's not a flake8 plugin we can use here, there is a tool called `pyupgrade` that we can use to make changes to the source, and it even exits non-zero so we can later use it as another checker. For now, can you try to run `pyupgrade` on your branch (it's conda-installable) and see what it finds? I was able to invoke it like this:
```sh
pyupgrade --py3-plus tests/**/*.py src/**/*.py examples/**/*.py docs/conf.py
```

dopplershift commented on .travis.yml:
```diff
@@ -38,8 +37,7 @@ env:
     # Codecov only works with coverage < 4.4
-    # Docutils 0.15 breaks sphinx 1.8.5 on Python 2.7
-    - EXTRA_PACKAGES="shapely<1.5.17.post1 coverage<4.4 docutils<0.15"
+    - EXTRA_PACKAGES="shapely<1.5.17.post1 coverage<4.4 docutils"
```
Should be able to just remove docutils here now--listing the version spec was just to hack around a problem on Python 2.

dopplershift commented on src/metpy/__init__.py:
```diff
@@ -4,21 +4,32 @@
+import sys
 import warnings
 
+if sys.version_info < (3,):
```
This is going to bug me, but it's part of the recommendations and we should probably do it. Annoying because this code will be flagged as having no coverage.

zbruick replied:
Agreed.

dopplershift commented:
Needs a rebase anyway now that I merged the 0.12 removal of deprecated stuff.

dopplershift commented:
Power-cycling to hopefully get codecov to actually track results--really want to see if there are any newly uncovered lines after this.
